### PR TITLE
Feature model training

### PR DESCRIPTION
**Pull request by anaston**
This new feature updates how the linear model is fitted on the reservoir states.

For instance, it allows to specify a model (e.g. SVM or logistic regression) instead of using ridge regression. It keeps ridge regression and all previous implementations as a deafult option (when model=None), however, I will point out in some comments that those implementations suffer from some inconsistences and have some issues.

Importantly, I introduce two new functions that are useful in classification tasks (see Lukosevicius (2012): A Practical Guide to Applying Echo State Networks. In: Montavon, Orr, Muller, editors. Lecture Notes in Computer Science. Springer, Berlin, Heidelberg, pp 659-686.)

- weighting readout reservoir states (i.e. inputs to the linear model) to assign different importance of time steps
- time averaging readout reservoir states within time-windows

Moreover, I fixed the issue with w_in which simply added the input signals when calculating the states of the reservoir nodes. Instead I recommend using an eye matrix (or diagonal matrix in general) to pass the input signals to different input nodes.

### CODE REVIEW

File: examples/esn.py [hunk 1]
```diff
@@ -17,79 +23,85 @@
...
+input_nodes = np.random.choice(subctx_nodes, n_features)
+output_nodes = np.where(ctx == 1)[0]  # we use cortical regions as output nodes
 
 # create input connectivity matrix, which defines the connec-
 # tions between the input layer (source nodes where the input signal is
 # coming from) and the input nodes of the reservoir.
 w_in = np.zeros((n_features, n_reservoir_nodes))
-w_in[np.ix_(np.arange(n_features), input_nodes)] = 10.0 # factor that modulates the activation state of the reservoir
+# factor that modulates the activation state of the reservoir
+w_in[np.ix_(np.arange(n_features), input_nodes)] = 10.0 * np.eye(n_features)
```
Comment (anaston) [inline examples/esn.py:87]: This fix allows to calculate the difference between two input signals which is essential, for instance, in the PerceptualDecisionMaking task. Before the fix, the input signals were summed up when calculating that state of the input nodes so the RC no longer has access to the input signals separately to be able to perform the task.

File: conn2res/iodata.py [hunk 2]
```diff
@@ -163,14 +176,13 @@ def fetch_dataset(task, n_trials=100, *args, **kwargs):
...
-    
-    return x, y
+        x = np.random.uniform(-1, 1, (1000+tau_max))[:, np.newaxis]
+        y = np.hstack([x[tau_max-tau:-tau] for tau in range(1, tau_max+1)])
 
+    return x, y
 
-def split_dataset(data, frac_train=0.7):
 
+def split_dataset(*args, frac_train=0.7, axis=0):
```
Comment (anaston) [inline conn2res/iodata.py:185]: I updated this function so that I can splits datasets in parallel, i.e. it can operate with multiple number of input data (e.g. x, y passed as *args).

Also, I defined an axis input argument as in some cases after data splitting we might need to concatenate along the second axis (i.e. axis=1).

File: conn2res/task.py [hunk 3]
```diff
@@ -19,77 +19,120 @@
...
+
     x_train, x_test = x
     y_train, y_test = y
 
-    model = Ridge(fit_intercept=False, alpha=0.5, **kwargs).fit(x_train, y_train)
-    score = model.score(x_test, y_test)
+    # specify default model
+    if model is None:
+        model = Ridge(fit_intercept=False, alpha=0.5,
+                      **kwargs)
```
Comment (anaston) [inline conn2res/task.py:62]: This default model that I kept for backward compatibility is used inconsistently:

- alpha is set to 0.5 whereas in other instances of ridge regression/classifier alpha is set to 0.0 or 1.0.

File: conn2res/task.py [hunk 4]
```diff
@@ -19,77 +19,120 @@
...
 
-    y_pred = model.predict(x_test)
-    n_outputs = y_pred.shape[1]
+        # calculate model scores on test data
+        y_pred = model.predict(x_test)
+        n_outputs = y_pred.shape[1]
+        score = []
+        for output in range(n_outputs):
+            score.append(
+                np.abs((np.corrcoef(y_test[:, output], y_pred[:, output])[0][1])))
```
Comment (anaston) [inline conn2res/task.py:101]: I kept this chunk of code as default for backward compatibility. However I would like to point out that the calculation of score in ridge regression is rather different from the implementation here (i.e. calculating the avarege correlation between the predicted and true outputs), so again it is inconsistent with other cases where you use calculate score of the ridge regression/classifier model.

File: conn2res/task.py [hunk 5]
```diff
@@ -19,77 +19,120 @@
...
+        'sample_weight', (None, None))
+
     x_train, x_test = x
     y_train, y_test = y
 
-    model = MultiOutputRegressor(Ridge(fit_intercept=False, alpha=0.5, **kwargs)).fit(x_train, y_train)
+    # specify default model
+    if model is None:
+        model = MultiOutputRegressor(
+            Ridge(fit_intercept=False, alpha=0.5, **kwargs))
```
Comment (anaston) [inline conn2res/task.py:89]: As we discussed, MultiOutputRegressor is actually not needed when used with Ridge as Ridge has the capability to perform multi-output regression off the shelf and gives the same result.

File: conn2res/task.py [hunk 6]
```diff
@@ -99,21 +142,40 @@ def classification(x, y, **kwargs):
...
     x_train, x_test = x
     y_train, y_test = y
 
-    # capture only decision time points
-    idx_train = np.nonzero(y_train)
-    idx_test  = np.nonzero(y_test)
+    # specify default model
+    if model is None:
+        model = OneVsRestClassifier(RidgeClassifier(
+            alpha=0.0, fit_intercept=False, **kwargs))
```
Comment (anaston) [inline conn2res/task.py:161]: Again, as we discussed in our meeting, OneVsRestClassifier is actually not needed when used with RidgeClassifier as RidgeClassifier has the capability to perform multi-class classification off the shelf and gives the same result.

Another drawback of using OneVsRestClassifier is that is does not support sample weights. In contrast, had we used RidgeClassifier, it supports sample weights.

File: conn2res/task.py [hunk 7]
```diff
@@ -127,47 +189,66 @@ def multiClassClassification(x, y, **kwargs):
...
     x_train, x_test = x
     y_train, y_test = y
 
-    model = MultiOutputClassifier(RidgeClassifier(alpha=0.5, fit_intercept=True, **kwargs)).fit(x_train, y_train)
-    score = model.score(x_test, y_test)    
-    
+    # specify default model
+    if model is None:
+        model = MultiOutputClassifier(RidgeClassifier(
+            alpha=0.5, fit_intercept=True, **kwargs))
```
Comment (anaston) [inline conn2res/task.py:208]: Again, as we discussed in our meeting, MultiOutputClassifier is actually not needed when used with RidgeClassifier as RidgeClassifier has the capability to perform multi-output classification off the shelf and gives the same result.

Another drawback of using MultiOutputClassifier is that is does not support sample weights. In contrast, had we used RidgeClassifier, it supports sample weights.

File: conn2res/iodata.py [hunk 8]
```diff
@@ -190,67 +204,122 @@ def split_dataset(data, frac_train=0.7):
...
-        
-        sns.despine(offset=10, trim=True)
-        # fig.savefig(fname=f'figs/{task}_io.png', transparent=True, bbox_inches='tight', dpi=300)
-        plt.show()
-        plt.close()
-    
+        plotting.plot_task(x, y, task)
+
+
+def get_sample_weight(inputs, labels, sample_block=None):
```
Comment (anaston) [inline conn2res/iodata.py:278]: Sorry, the docstring of the function is outdated. This function finds blocks in the data based on repetition in the signal then it defines a sample_weight proportional to the length of this block. For instance, the signal [1 1 1 1 2] would produce sample_weight [0.25 0.25 0.25 0.25 1].

File: conn2res/task.py [hunk 9]
```diff
@@ -99,21 +142,40 @@ def classification(x, y, **kwargs):
...
+    if isinstance(model, OneVsRestClassifier):
+        # select decision time points
+        idx_train = np.nonzero(y_train)
+        idx_test = np.nonzero(y_test)
+
+        # fit model on training data (OneVsRestClassifier does not support sample weights)
+        model.fit(x_train[idx_train], y_train[idx_train])
+
+        # calculate model scores on test data
+        score = model.score(x_test[idx_test], y_test[idx_test])
```
Comment (anaston) [inline conn2res/task.py:172]: I kept the usage of idx_train and idx_test for backward compatibility. However, as we discussed in our meeting, it makes more sense to train on all data and test as decision time points.

With the new sample_weight_test functionality, the decision time points can be defined there as a Boolean, i.e. passing weight=1 for at decision time points and weight=0 elsewhere.

File: conn2res/coding.py [hunk 10]
```diff
@@ -65,42 +68,104 @@ def encoder(reservoir_states, target, readout_modules=None,\
...
-    
+
     else:
         df_encoding = run_task(reservoir_states=reservoir_states,
                                target=target, **kwargs)
 
     return df_encoding
+
+
+def time_average_samples(seq_len, data, sample_weight, operation=None):
```
Comment (anaston) [inline conn2res/coding.py:113]: This function averages reservoir states (inputs to the linear model) or subsamples labels (outputs of the linear model) to match the size of the reservoir states after averaging. In both cases, it uses sample weights to achieve the desired result.
